File: backend/src/services/impactScore.ts
import type { Intervention, ImpactBreakdown } from '../types';

const VEG_WEIGHT = 0.30;
const WATER_WEIGHT = 0.25;
const LAND_WEIGHT = 0.15;
const STRUCTURE_WEIGHT = 0.15;
const HYDRO_WEIGHT = 0.10;
const RAIN_WEIGHT = 0.05;

function normalizeChange(before: number, after: number, maxRange = 1): number {
  const delta = after - before;
  const norm = Math.max(-1, Math.min(1, delta / maxRange));
  const score = 50 + norm * 50;
  return Math.max(0, Math.min(100, score));
}

function conditionScore(condition: Intervention['condition']): number {
  switch (condition) {
    case 'Excellent': return 95;
    case 'Good': return 80;
    case 'Moderate': return 55;
    case 'Poor': return 30;
    default: return 50;
  }
}

function hydroRelevance(int: Intervention): number {
  const waterTypes = ['Check Dam', 'Farm Pond'];
  if (!waterTypes.includes(int.type)) return 65;
  if (int.waterAreaAfter > 3) return 92;
  if (int.waterAreaAfter > 1.5) return 75;
  if (int.waterPresent) return 55;
  return 35;
}

function rainfallAdjusted(int: Intervention): number {
  const r = int.rainfallChange;
  if (r >= 15) return 90;
  if (r >= 8) return 78;
  if (r >= 2) return 62;
  if (r >= -4) return 48;
  return 30;
}

export function calculateImpactScore(intervention: Intervention): ImpactBreakdown {
  const vegetation = normalizeChange(intervention.ndviBefore, intervention.ndviAfter, 0.5);
  const water = normalizeChange(intervention.ndwiBefore, intervention.ndwiAfter, 0.3) * 0.5 +
    normalizeChange(intervention.waterAreaBefore, intervention.waterAreaAfter, 5) * 0.5;
  const landCover = (vegetation + water) / 2;
  const structure = conditionScore(intervention.condition);
  const hydrological = hydroRelevance(intervention);
  const rainfallAdjustedScore = rainfallAdjusted(intervention);

  const score =
    vegetation * VEG_WEIGHT +
    water * WATER_WEIGHT +
    landCover * LAND_WEIGHT +
    structure * STRUCTURE_WEIGHT +
    hydrological * HYDRO_WEIGHT +
    rainfallAdjustedScore * RAIN_WEIGHT;

  const finalScore = Math.round(Math.max(0, Math.min(100, score)));

  let label = 'Moderate Impact';
  if (finalScore >= 75) label = 'High Impact';
  else if (finalScore >= 50) label = 'Moderate Impact';
  else if (finalScore >= 35) label = 'Needs Review';
  else label = 'At Risk';

  return {
    score: finalScore,
    vegetation: Math.round(vegetation),
    water: Math.round(water),
    landCover: Math.round(landCover),
    structure: Math.round(structure),
    hydrological: Math.round(hydrological),
    rainfallAdjusted: Math.round(rainfallAdjustedScore),
    label,
  };
}

export function riskLabel(score: number): 'High Impact' | 'Moderate Impact' | 'At Risk' | 'Needs Review' {
  if (score >= 75) return 'High Impact';
  if (score >= 50) return 'Moderate Impact';
  if (score >= 35) return 'Needs Review';
  return 'At Risk';
}

export function riskColor(score: number): string {
  if (score >= 75) return '#22c55e';
  if (score >= 50) return '#f59e0b';
  if (score >= 35) return '#f97316';
  return '#ef4444';
}

File: backend/app/services/impact_score.py
"""
Impact score + recommendation engine.

BUG FIXES vs TypeScript services/impactScore.ts:
- ImpactBreakdown.label now returns the NARROW RiskLevel literal (not `str`),
  fixing the buggy cast `as Intervention['riskLevel']` in the TS code.
- Weights sum to 1.0 exactly (VEG + WATER + LAND + STRUCT + HYDRO + RAIN = 1.0).
- `risk_label()` exposes the deterministic mapping for direct reuse.
"""
from __future__ import annotations

from dataclasses import dataclass

from ..schemas import (
    ImpactBreakdownSchema,
    InterventionBase,
    RiskLevel,
)


VEG_WEIGHT = 0.30
WATER_WEIGHT = 0.25
LAND_WEIGHT = 0.15
STRUCTURE_WEIGHT = 0.15
HYDRO_WEIGHT = 0.10
RAIN_WEIGHT = 0.05


def _normalize_change(before: float, after: float, max_range: float = 1.0) -> float:
    delta = after - before
    norm = max(-1.0, min(1.0, delta / max_range))
    return max(0.0, min(100.0, 50.0 + norm * 50.0))


def _condition_score(condition: str) -> float:
    return {
        "Excellent": 95.0,
        "Good": 80.0,
        "Moderate": 55.0,
        "Poor": 30.0,
    }.get(condition, 50.0)


def _hydro_relevance_score(intv: InterventionBase) -> float:
    water_types = {"Check Dam", "Farm Pond"}
    if intv.type not in water_types:
        return 65.0
    if intv.waterAreaAfter > 3:
        return 92.0
    if intv.waterAreaAfter > 1.5:
        return 75.0
    if intv.waterPresent:
        return 55.0
    return 35.0


def _rainfall_adjusted_score(intv: InterventionBase) -> float:
    r = intv.rainfallChange
    if r >= 15:
        return 90.0
    if r >= 8:
        return 78.0
    if r >= 2:
        return 62.0
    if r >= -4:
        return 48.0
    return 30.0


def risk_label(score: int) -> RiskLevel:
    """Deterministic label from a 0..100 impact score. Mirrors frontend thresholds."""
    if score >= 75:
        return "High Impact"
    if score >= 50:
        return "Moderate Impact"
    if score >= 35:
        return "Needs Review"
    return "At Risk"


def risk_color(score: int) -> str:
    return {
        "High Impact": "#22c55e",
        "Moderate Impact": "#f59e0b",
        "Needs Review": "#f97316",
        "At Risk": "#ef4444",
    }[risk_label(score)]


def calculate_impact_score(intervention: InterventionBase) -> ImpactBreakdownSchema:
    vegetation = _normalize_change(intervention.ndviBefore, intervention.ndviAfter, max_range=0.5)
    water = (
        0.5 * _normalize_change(intervention.ndwiBefore, intervention.ndwiAfter, max_range=0.3)
        + 0.5 * _normalize_change(intervention.waterAreaBefore, intervention.waterAreaAfter, max_range=5.0)
    )
    land_cover = (vegetation + water) / 2.0
    structure = _condition_score(intervention.condition)
    hydro = _hydro_relevance_score(intervention)
    rainfall = _rainfall_adjusted_score(intervention)

    score = (
        vegetation * VEG_WEIGHT
        + water * WATER_WEIGHT
        + land_cover * LAND_WEIGHT
        + structure * STRUCTURE_WEIGHT
        + hydro * HYDRO_WEIGHT
        + rainfall * RAIN_WEIGHT
    )
    final = int(round(max(0.0, min(100.0, score))))

    return ImpactBreakdownSchema(
        score=final,
        vegetation=int(round(vegetation)),
        water=int(round(water)),
        landCover=int(round(land_cover)),
        structure=int(round(structure)),
        hydrological=int(round(hydro)),
        rainfallAdjusted=int(round(rainfall)),
        label=risk_label(final),
    )
